### PR TITLE
#2294 Making Spinners great again

### LINKED ISSUE #2294: Spinner when creating program requisition
## Is your feature request related to a problem? Please describe.

With newly added logic for program requisition SOH and AMC calculation, need an indicator that requisition is loading (as it takes a bit longer and looks like app frozen)

## Describe the solution you'd like

Have the `runWithLoadingIndicator` spinner display when the requisition is being navigated to.

## Implementation

Wrap the call to navigate in `runWithLoadingIndicator`. To access the function you can use `LoadingIndicatorContext`

## Describe alternatives you've considered

n/a

## Additional context

n/a

### PR DESCRIPTION
**Pull request by joshxg**
Fixes #2294

## Change summary

- Makes Spinners work for supplierREquisition creating from program and adding from master lists
- Makes spinners work for adding items from master list for customer invoice page
- Makes spinners work for creating a stocktake
- Fixes a type bug in customer invoice `mapDispatchToProps`

## Testing

- [ ] When creating a `SupplierRequisition` from a program, a spinner is displayed.
- [ ] When adding items from a/many master lists to a supplierRequisition a spinner is displayed.
- [ ] When adding items from a/many master lists to a customerInvoice a spinner is displayed.
- [ ] When creating a `stocktake` from a program, a spinner is displayed.

### Related areas to think about

N/A

### CODE REVIEW

File: src/actions/FinaliseActions.js [hunk 1]
```diff
@@ -40,10 +39,7 @@ const finalise = () => (dispatch, getState) => {
     bugsnagClient.notify(error);
   }
 
-  batch(() => {
-    dispatch(closeModal());
-    dispatch(refreshData(currentRoute));
-  });
+  dispatch(refreshData(currentRoute));
```
Comment (joshxg) [inline src/actions/FinaliseActions.js:42]: delegating `closeModal` now to the caller

File: src/mSupplyMobileApp.js [hunk 2]
```diff
@@ -143,7 +143,8 @@ class MSupplyMobileAppContainer extends React.Component {
     await new Promise(resolve => {
       this.setState({ isLoading: true }, () => setTimeout(resolve, 1));
     });
-    functionToRun();
+
+    await functionToRun();
```
Comment (joshxg) [inline src/mSupplyMobileApp.js:147]: awaiting a non-promise converts it to a promise, so this is OK, apparently!

File: src/pages/CustomerInvoicePage.js [hunk 3]
```diff
@@ -75,6 +76,13 @@ export const CustomerInvoice = ({
     isFinalised,
   ]);
 
+  const runWithLoadingIndicator = useLoadingIndicator();
+
+  const onAddFromMasterLists = selected => {
+    onCloseModal();
+    runWithLoadingIndicator(() => onApplyMasterLists(selected, pageObject));
+  };
```
Comment (joshxg) [inline src/pages/CustomerInvoicePage.js:84]: pattern to be used in the rest of the PR: Close the modal before starting the spinner as that will more or less block the JS thread and it won't close the modal to show the spinner

File: src/pages/CustomerInvoicePage.js [hunk 4]
```diff
@@ -200,10 +208,15 @@ export const CustomerInvoice = ({
 };
 
 const mapDispatchToProps = (dispatch, ownProps) => {
-  const { transaction } = ownProps?.navigation?.route?.param || {};
+  const { transaction } = ownProps?.route?.params || {};
```
Comment (joshxg) [inline src/pages/CustomerInvoicePage.js:211]: bug fix - `params` is correct

File: src/mSupplyMobileApp.js (hunk 2 above)
Comment (wlthomson) [reply src/mSupplyMobileApp.js:147]: If it works, it works!

File: src/pages/CustomerInvoicePage.js (hunk 3 above)
Comment (wlthomson) [reply src/pages/CustomerInvoicePage.js:84]: thumbsup

File: src/pages/CustomerInvoicePage.js (hunk 4 above)
Comment (wlthomson) [reply src/pages/CustomerInvoicePage.js:211]: Nit: `ownProps?.route?.params ?? {}`.

**Review by wlthomson — APPROVED** [review]
lgtm

**Pull request merged**